app/feature/feature-movingflow/src/main/kotlin/com/hedvig/android/feature/movingflow/ui/selectcontract/SelectContractViewModel.kt

```diff
@@ -0,0 +1,189 @@
+          val state = currentState as? Content ?: return@CollectEvents
+          currentState = state.copy(navigateToHousingType = false)
+        }
+
+        is SelectContractEvent.SelectContract -> {
+          val state = currentState as? Content ?: return@CollectEvents
+          currentState = state.copy(
+            selectedAddress = event.contract,
+          )
+        }
+
+        SelectContractEvent.SubmitContract -> {
+          val state = currentState as? Content ?: return@CollectEvents
+          val selectedId = state.selectedAddress ?: return@CollectEvents
+          submittingAddressId = selectedId.id
+        }
+
+        SelectContractEvent.RetryLoadData -> loadIteration++
+      }
+    }
+
+    LaunchedEffect(submittingAddressId) {
+      val id = submittingAddressId
+      if (id != null) {
+        val state = currentState as? NotEmpty ?: return@LaunchedEffect
+        if (state is Content) {
+          currentState = state.copy(buttonLoading = true)
+        }
+        val moveIntent = state.intent
+        movingFlowRepository.initiateNewMovingFlow(moveIntent, id)
+        submittingAddressId = null
+        currentState = when (state) {
+          is Content -> state.copy(navigateToHousingType = true, buttonLoading = false)
+          is Redirecting -> state.copy(navigateToHousingType = true)
+        }
+      }
+    }
+
+    LaunchedEffect(loadIteration) {
+      either {
+        val moveIntentCreate = apolloClient
+          .mutation(MoveIntentV2CreateMutation())
+          .safeExecute()
+          .mapLeft(::ErrorMessage)
+          .mapLeft { SelectContractState.Error.GenericError(it) }
+          .map { it.moveIntentCreate }
+          .bind()
+        val moveIntent = ensureNotNull(moveIntentCreate.moveIntent) {
+          val userError = moveIntentCreate.userError?.message
+          if (userError == null) {
+            SelectContractState.Error.GenericError(
+              ErrorMessage(
+                "Unknown MoveIntentV2CreateMutation error",
+              ),
+            )
+          } else {
+            SelectContractState.Error.UserPresentable(userError)
+          }
+        }
+        moveIntent
+      }.fold(
+        ifLeft = { error ->
+          Snapshot.withMutableSnapshot {
+            currentState = error
+          }
+        },
+        ifRight = { intent ->
+          Snapshot.withMutableSnapshot {
+            val currentAddressesSize = intent.currentHomeAddresses.size
+            when {
+              currentAddressesSize > 1 -> {
+                currentState = Content(
+                  intent = intent,
+                  selectedAddress = null,
+                  navigateToHousingType = false,
+                  buttonLoading = false,
+                )
+              }
+
+              currentAddressesSize == 1 -> {
+                submittingAddressId = intent.currentHomeAddresses[0].id
+                currentState = Redirecting(
```

Comment:
^^ That was a lie.
It seems like here we do set two variables together. I'd suggest we just wrap these two with the `Snapshot.withMutableSnapshot` here so that we make it more obvious why we use it here 😊